util/network-devp2p/src/discovery.rs

```diff
@@ -233,17 +249,17 @@ impl Discovery {
 		self.discovery_round += 1;
 	}
 
-	fn distance(a: &H256, b: &H256) -> u32 {
-		let d = *a ^ *b;
-		let mut ret:u32 = 0;
-		for i in 0..32 {
-			let mut v: u8 = d[i];
-			while v != 0 {
-				v >>= 1;
-				ret += 1;
+	/// The base 2 log of the distance between a and b using the XOR metric.
+	fn distance(a: &H256, b: &H256) -> Option<usize> {
+		for i in (0..ADDRESS_BYTES_SIZE).rev() {
+			let byte_index = ADDRESS_BYTES_SIZE - i - 1;
+			let d: u8 = a[byte_index] ^ b[byte_index];
+			if d != 0 {
```

Comment:
Do we need to deal with timing attack here because I suppose this particular function is a crypto function? In the case where network connection is really stable, would it be possible for an attacker to conduct timing attack and leak some information through `PING` time? That function calls `update_node` and uses `distance` when inserting.

Don't know whether that's actually feasible, but if so, maybe we can also just deal with `d != 0` case and return it after 32 loops.

Reply:
To elaborate more on how this may be an issue (while a really subtle one), an attacker can use timing attack through `PING` on a stable network to guess the bucket index in a range of 8 (because of early return). Then, even if we applied eclipse attack countermeasure 3, the bucket index is still possible to be guessed with an offset of 8.

Public bucket index is an attack factor in the eclipse attack, so it may be possible to then use this information to conduct some sort of similar attacks.

Reply:
Yes, as long as the distance metric is public, I don't see the risk. If the distance is salted as recommended by the paper, this could be easily defended by sending the pong response packet before calling `update_node` (which would probably be a good idea regardless). Also, I generally think salting the distance is a bad idea...